core/src/main/java/com/datastax/oss/driver/api/core/data/GettableByIndex.java

```diff
@@ -427,10 +469,13 @@ default Token getToken(int i) {
    * <p>This method is provided for convenience when the element type is a non-generic type. For
    * more complex list types, use {@link #get(int, GenericType)}.
    *
+   * @return the {@code i}th value as a Java map; may be empty but never {@code null}.
    * @throws IndexOutOfBoundsException if the index is invalid.
    */
+  @Nonnull
   default <T> List<T> getList(int i, Class<T> elementsClass) {
-    return get(i, GenericType.listOf(elementsClass));
+    List<T> value = get(i, GenericType.listOf(elementsClass));
+    return value == null ? new ArrayList<>(0) : value;
```

Comment:
Hrmm, is this a hard requirement that it returns an empty list?  Shouldn't it be dependent on the registered codec?  The list codec will return nulls as empty collections, but I see that the 3.x code defers to the codec and doesn't make any assumptions otherwise:

https://github.com/datastax/java-driver/blob/3.x/driver-core/src/main/java/com/datastax/driver/core/GettableByIndexData.java#L306-L309

Reply:
We discussed yesterday about collections and I was under the impression that the general policy is to never return null, but rather empty collections for List, Set and Map. This change is required to comply with the `@Nonnull` annotations. But I don't have strong opinions here, changing the contract to `@Nullable` is fine too. Also, I'm returning regular java.util collections rather than immutable structures as I saw that the built-in codecs do the same.

Reply:
When we talked of empty collections, I was thinking more of session-level methods, such as `Metadata.getKeyspaces()`: return an empty collection even when schema metadata is disabled, instead of wrapping the collection in an `Optional`.

For row-level getters I think we should keep the current behavior of depending on the registered codec.

I kept the returned collections mutable because I felt it would trip users too much if you can't manipulate the result directly, and the risk of thread safety issues when you're dealing with a simple column value is pretty minimal.